src/main/kotlin/org/jetbrains/research/ml/codetracker/FileHandler.kt

```diff
@@ -0,0 +1,93 @@
+
+    init {
+        logger.info("${Plugin.PLUGIN_ID}: init documents listener")
+    }
+
+    // Tracking documents changes before to be consistent with activity-tracker plugin
+    override fun beforeDocumentChange(event: DocumentEvent) {
+        if (isValidChange(event)) DocumentLogger.log(event.document)
+    }
+
+    // To avoid completion events with IntellijIdeaRulezzz sign
+    // Todo: should we use it?
+    private fun isValidChange(event: DocumentEvent): Boolean {
+        return FileHandler.isTrackedDocument(event.document)
+                && EditorFactory.getInstance().getEditors(event.document).isNotEmpty()
+                && FileDocumentManager.getInstance().getFile(event.document) != null
+    }
+}
+
+object FileHandler {
+    private const val PLUGIN_FOLDER = "codetracker"
+
+    private val logger: Logger = Logger.getInstance(javaClass)
+    private val documentToTask: HashMap<Document, Task> = HashMap()
+
+    private val listener by lazy {
+        MyDocumentListener()
+    }
+
+    private fun createFile(project: Project, task: Task, language: Language): VirtualFile? {
+        val file = File("${project.basePath}/${PLUGIN_FOLDER}/${task.key}${language.extension.ext}")
+        FileUtil.createIfDoesntExist(file)
+        return LocalFileSystem.getInstance().refreshAndFindFileByIoFile(file)
+    }
+
+    // If the documentToTask has task null then we don't track the document
+    private fun addDocument(virtualFile: VirtualFile?, task: Task): Document? {
+        val document = virtualFile?.let { FileDocumentManager.getInstance().getDocument(it) }
+        if (!documentToTask.containsKey(document)) {
+            document?.let { documentToTask.put(it, task) }
+            document?.addDocumentListener(listener)
+        } else {
+            // It the old task is not equal the new task
+            if (documentToTask[document] != task) {
+                val message = "${Plugin.PLUGIN_ID}: an attempt to assign another task to the document ${document}. " +
+                        "The old task is ${documentToTask[document]!!.key}, the new task is ${task.key}"
+                logger.error(message)
+                throw IllegalArgumentException(message)
+            }
+        }
+        return document
+    }
+
+    fun createAndOpenFile(project: Project, task: Task, language: Language = Language.PYTHON): VirtualFile? {
+        val virtualFile = createFile(project, task, language)
+        addDocument(virtualFile, task)
+        openFile(project, virtualFile)
+        return virtualFile
+    }
+
+    fun isTrackedDocument(document: Document): Boolean {
+        return documentToTask.containsKey(document)
+    }
+
+    private fun openFile(project: Project, virtualFile: VirtualFile?) {
+        virtualFile?.let { FileEditorManager.getInstance(project).openFile(it, true, true) }
+    }
+
+    fun stopTracking() {
+        documentToTask.map { (d, _) -> d.removeDocumentListener(listener) }
```

Comment:
Shouldn't you use `forEach` instead of `map`? Also, you can use `it.key.removeDocumentListener(listener)`, it's a little more readable and short.